### PR TITLE
Added method to create a new domain object for the specified bounds and axes

### LINKED ISSUE #525: Function for creating a domain with a regular lat-lon grid
It is common to want to regrid a field to a new global grid with constant latitude-longitude cell sizes. In the case that the there is no "destination" field or domain to get this grid from, it is a bit fiddly to create it from scratch, involving the creation of `DimensionCoordinate` constructs and their bounds, and binding them all together.

It would be good to have a cf function that could produce such a domain as a one-liner, something like (function name and API for demo purposes only):
```python
>>> d = cf.create_global_grid(dlat=1, dlon=1, origin_lon=0)  # create a global 1 degree grid [0, 360], [-90, 90]
>>> d.dump()
--------
Domain: 
--------

Domain Axis: latitude(180)
Domain Axis: longitude(360)

Dimension coordinate: latitude
    standard_name = 'latitude'
    units = 'degrees_north'
    Data(latitude(180)) = [-89.5, ..., 89.5] degrees_north
    Bounds:units = 'degrees_north'
    Bounds:Data(latitude(180), 2) = [[-90.0, ..., 90.0]] degrees_north

Dimension coordinate: longitude
    standard_name = 'longitude'
    units = 'degrees_east'
    Data(longitude(360)) = [0.5, ..., 359.5] degrees_east
    Bounds:units = 'degrees_east'
    Bounds:Data(longitude(360), 2) = [[0.0, ..., 360.0]] degrees_east
```

Thanks to @CheeseFace for suggesting this feature

### PR DESCRIPTION
**Pull request by bewithankit**
Fixes #525

Added method to create a new ``Domain`` object with the specified longitude and latitude bounds, domain axes, and coordinate axes. An example use case would be using ``cf.Domain.create`` to define the destination grid while regridding, as suggested by @davidhassell.

### CODE REVIEW

File: cf/domain.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from math import prod
 
+import numpy
```
Comment (davidhassell) [inline cf/domain.py:3]:
```suggestion
import numpy as np
```

File: cf/domain.py [hunk 2]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+        x_range = x[1] - x[0]
+        y_range = y[1] - y[0]
+
+        if not (x_range % dx == 0 and y_range % dy == 0):
+            raise ValueError("dx and dy must be divisible by the respective ranges of x and y.")
+
+        domain = cls()
+
+        x_points = numpy.arange(x[0], x[1] + dx, dx)
+        y_points = numpy.arange(y[0], y[1] + dy, dy)
```
Comment (davidhassell) [inline cf/domain.py:207]:
```suggestion
        x_points = np.arange(x[0], x[1] + dx, dx)
        y_points = np.arange(y[0], y[1] + dy, dy)
```

'x_points` doesn't work for cyclic axes, for which we don't repeat include the last value
```python
 >>> np.arange(-180, 180 + 30, 30)
array([-180, -150, -120,  -90,  -60,  -30,    0,   30,   60,   90,  120,
        150,  180])
```
Decreasing ranges similarly:
```python
>>> np.arange(-180, 180 + -30, -30)
array([], dtype=int64)
```

That said, we need to think through what to do in this case - should `x` and `y` also be decreasing?

File: cf/domain.py [hunk 3]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
 
+    @classmethod
+    def create(cls, x, y, dx, dy):
+        '''
+        Create a new domain with the given longitude and latitude bounds and intervals.
+        
+        :Parameters:
+        
+            x: tuple
+                A tuple of two numeric values representing the lower and upper bounds of the longitude dimension.
```
Comment (davidhassell) [inline cf/domain.py:169]: Line wrapping

File: cf/domain.py [hunk 4]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+        
+            y: tuple
+                A tuple of two numeric values representing the lower and upper bounds of the latitude dimension.
```
Comment (davidhassell) [inline cf/domain.py:172]: Line wrapping

File: cf/domain.py [hunk 5]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+            dx: float
+                The interval between points along the longitude axis.
+        
+            dy: float
+                The interval between points along the latitude axis.
+        
+        :Returns:
+        
+            `Domain`
+                The newly created domain with the specified longitude and latitude bounds, domain axes, and coordinate axes.
```
Comment (davidhassell) [inline cf/domain.py:183]: Line wrapping

File: cf/domain.py [hunk 6]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+        
+        **Examples:**
+        
+        >>> import cf
+        >>> domain = cf.Domain.create((-180, 180), (-90, 90), 1, 1)
+        >>> print(domain)
```
Comment (davidhassell) [inline cf/domain.py:189]: May be better to do `domain.dump()`, as it will also show the bounds.

File: cf/domain.py [hunk 7]
```diff
@@ -156,6 +158,79 @@ def close(self):
             removed_at="5.0.0",
         )  # pragma: no cover
 
+    @classmethod
+    def create(cls, x, y, dx, dy):
+        '''
+        Create a new domain with the given longitude and latitude bounds and intervals.
+
```
Comment (davidhassell) [inline cf/domain.py:165]:
```suggestion

        .. versionadded:: 3.15.0

```

File: cf/domain.py [hunk 8]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+            raise ValueError("dx and dy must be divisible by the respective ranges of x and y.")
+
+        domain = cls()
+
+        x_points = numpy.arange(x[0], x[1] + dx, dx)
+        y_points = numpy.arange(y[0], y[1] + dy, dy)
+
+        # Create domain axes for longitude and latitude
+        domain_axis_longitude = domain.set_construct(domain._DomainAxis(len(x_points)))
+        domain_axis_latitude = domain.set_construct(domain._DomainAxis(len(y_points)))
```
Comment (davidhassell) [inline cf/domain.py:211]:
```suggestion
        domain_axis_longitude = domain.set_construct(domain._DomainAxis(x_points.size))
        domain_axis_latitude = domain.set_construct(domain._DomainAxis(y_points.size))
```
`.size` is faster :)

File: cf/domain.py [hunk 9]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+
+        if not (len(x) == 2 and len(y) == 2):
+            raise ValueError("x and y must have two elements each.")
+
+        x_range = x[1] - x[0]
+        y_range = y[1] - y[0]
+
+        if not (x_range % dx == 0 and y_range % dy == 0):
+            raise ValueError("dx and dy must be divisible by the respective ranges of x and y.")
+
```
Comment (davidhassell) [inline cf/domain.py:203]: I think that we should test on `x_range` and `y_range` being positive, too (see below).

File: cf/domain.py [hunk 10]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+        '''
+
+        if not (len(x) == 2 and len(y) == 2):
+            raise ValueError("x and y must have two elements each.")
+
+        x_range = x[1] - x[0]
+        y_range = y[1] - y[0]
+
+        if not (x_range % dx == 0 and y_range % dy == 0):
+            raise ValueError("dx and dy must be divisible by the respective ranges of x and y.")
```
Comment (davidhassell) [inline cf/domain.py:202]: Don't forget to run `pre-commit run --all-files`.

File: cf/domain.py [hunk 11]
```diff
@@ -156,6 +158,79 @@ def close(self):
...
+                A tuple of two numeric values representing the lower and upper bounds of the longitude dimension.
+        
+            y: tuple
+                A tuple of two numeric values representing the lower and upper bounds of the latitude dimension.
+        
+            dx: float
+                The interval between points along the longitude axis.
+        
+            dy: float
+                The interval between points along the latitude axis.
```
Comment (davidhassell) [inline cf/domain.py:178]: Need some info on what happens if dy < 0

**Review by davidhassell — CHANGES_REQUESTED** [review]
Hi Ankit,

Thanks for the quick PR!

Perhaps not a complete review, but a few things to be get us started.

File: cf/domain.py (hunk 11 above)
Comment (bewithankit) [reply cf/domain.py:183]: True, the array will be empty. Should I add a check for negative `dy` and reverse the order of the array?
```python
if dy < 0:
    y_points = np.flip(y_points)

File: cf/dimensioncoordinate.py [hunk 12]
```diff
@@ -334,6 +336,98 @@ def direction(self):
 
         return _direction
 
+    @classmethod
+    def createregular(
+        cls, range, interval, use_bounds=True, units=None, standard_name=None
+    ):
+        """
+        Create a new DimensionCoordinate with the given range and interval.
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:344]:
```suggestion
        Create a new `DimensionCoordinate` with the given range and regular cell sizes.
```

File: cf/domain.py [hunk 13]
```diff
@@ -1,6 +1,9 @@
 from math import prod
 
 import cfdm
+import numpy as np
+
+import cf
```
Comment (davidhassell) [inline cf/domain.py:6]:
```suggestion
import cf
```
Why do we need to import this?

File: cf/domain.py [hunk 14]
```diff
@@ -156,6 +159,94 @@ def close(self):
             removed_at="5.0.0",
         )  # pragma: no cover
 
+    @classmethod
+    def create(cls, x, y, dx, dy, use_bounds=True):
```
Comment (davidhassell) [inline cf/domain.py:163]: Perhaps call this `create_regular`, too?

File: cf/domain.py [hunk 15]
```diff
@@ -156,6 +159,94 @@ def close(self):
...
+            units = 'degrees_north'
+            Data(latitude(180)) = [-89.5, ..., 89.5] degrees_north
+            Bounds:units = 'degrees_north'
+            Bounds:Data(latitude(180), 2) = [[-90.0, ..., 90.0]] degrees_north
+
+
+        """
+        domain = cls()
+
+        longitude = DimensionCoordinate.createregular(
```
Comment (davidhassell) [inline cf/domain.py:229]:
```suggestion
        longitude = domain._DimensionCoordinate.createregular(
```

File: cf/domain.py [hunk 16]
```diff
@@ -156,6 +159,94 @@ def close(self):
...
+            x, dx, use_bounds, "degrees_east", "longitude"
+        )
+
+        latitude = DimensionCoordinate.createregular(
```
Comment (davidhassell) [inline cf/domain.py:233]:
```suggestion
        latitude = domain._DimensionCoordinate.createregular(
```

File: cf/domain.py [hunk 17]
```diff
@@ -156,6 +159,94 @@ def close(self):
...
+    def create(cls, x, y, dx, dy, use_bounds=True):
+        """
+        Create a new domain with the given longitude and latitude bounds and
+        intervals.
+
+        .. versionadded:: 3.15.0
+
+        :Parameters:
+
+            x: tuple
```
Comment (davidhassell) [inline cf/domain.py:172]:
```suggestion
            x_range: `tuple`
```

File: cf/domain.py [hunk 18]
```diff
@@ -156,6 +159,94 @@ def close(self):
...
+                A tuple of two numeric values representing the lower and
+            upper bounds of the longitude dimension.
+
+            y: tuple
+                A tuple of two numeric values representing the lower and
+            upper bounds of the latitude dimension.
+
+            dx: float
```
Comment (davidhassell) [inline cf/domain.py:180]:
```suggestion
            dx: numerical scalar or `Timeduration`
```

**Review by davidhassell — CHANGES_REQUESTED** [review]
Hi Ankit - an interim review. I'll post more tomorrow. Thanks

File: cf/domain.py [hunk 19]
```diff
@@ -330,6 +330,111 @@ def del_file_location(
...
         return location
 
+    @classmethod
+    def create_regular(cls, x_args, y_args, bounds=True):
+        """
+        Create a new domain with the given longitude and latitude bounds and
+        intervals.
+
+        .. versionadded:: 3.15.1
+
```
Comment (davidhassell) [inline cf/domain.py:340]:
```suggestion

        .. seealso:: `cf.DimensionCoordinate.create_regular`
        
```

File: cf/dimensioncoordinate.py [hunk 20]
```diff
@@ -334,6 +334,106 @@ def direction(self):
 
         return _direction
 
+    @classmethod
+    def create_regular(
+        cls, args, units=None, standard_name=None, bounds=True
+    ):
+        """
+        Create a new `DimensionCoordinate` with the given range and regular
+        cellsize.
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:343]: PEP8 tells us to keep this to one line. How about
```suggestion
        Create a new `DimensionCoordinate` with the given range and cellsize.
```

File: cf/dimensioncoordinate.py [hunk 21]
```diff
@@ -334,6 +334,106 @@ def direction(self):
...
+
+        .. versionadded:: 3.15.0
+
+        :Note: This method does not set the cyclicity of the
+         `DimensionCoordinate`.
+
+        :Parameters:
+
+            args: `sequence of numbers`
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:352]:
```suggestion
            args: sequence of numbers
```

File: cf/dimensioncoordinate.py [hunk 22]
```diff
@@ -334,6 +334,106 @@ def direction(self):
...
+                A sequence of three numeric values representing the lower
+                and upper bounds of the dimension and the cellsize.
+                The first two values in the sequence represent the range,
+                and the third value represents the cellsize.
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:356]: Perhaps reword this to not use the word "bounds", as that could be misleading

"A sequence of three numeric values. The first two values in the sequence represent the coordinate range (see the *bounds* parameter for details),  and the third value represents the cellsize."

File: cf/domain.py [hunk 23]
```diff
@@ -330,6 +330,111 @@ def del_file_location(
...
+        :Parameters:
+
+            x_args: `sequence of numbers`
+                A sequence of three numeric values representing the lower and
+                upper bounds of the longitude dimension and the interval between
+                points along the longitude axis.
```
Comment (davidhassell) [inline cf/domain.py:346]: Perhaps reword this to not use the word "bounds", as that could be misleading

"A sequence of three numeric values. The first two values in the sequence represent the coordinate range (see the bounds parameter for details), and the third value represents the cellsize."

File: cf/domain.py [hunk 24]
```diff
@@ -330,6 +330,111 @@ def del_file_location(
...
+
+            y_args: `sequence of numbers`
+                A sequence of three numeric values representing the lower and
+                upper bounds of the latitude dimension and the interval between
+                points along the latitude axis.
```
Comment (davidhassell) [inline cf/domain.py:351]: Perhaps reword this to not use the word "bounds", as that could be misleading

"A sequence of three numeric values. The first two values in the sequence represent the coordinate range (see the bounds parameter for details), and the third value represents the cellsize."

**Review by davidhassell — CHANGES_REQUESTED** [review]

File: cf/dimensioncoordinate.py [hunk 25]
```diff
@@ -334,6 +334,111 @@ def direction(self):
...
+        :Parameters:
+
+            args: sequence of numbers
+                A sequence of three numeric values. The first two values in the 
+                sequence represent the coordinate range (see the bounds 
+                parameter for details), and the third value represents the 
+                cellsize.
+
+            bounds: `bool`, optional
+                If True (default), the given range represents
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:359]:
```suggestion
                If True (the default) then the given range represents
```

File: cf/dimensioncoordinate.py [hunk 26]
```diff
@@ -334,6 +336,98 @@ def direction(self):
...
+        if range_diff % interval != 0:
+            raise ValueError(
+                "interval must be divisible by the range of the dimension."
+            )
+
+        if not isinstance(use_bounds, bool):
+            raise ValueError("use_bounds must be a boolean.")
+
+        if units is not None and not isinstance(units, str):
+            raise ValueError("units must be either None or a string.")
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/dimensioncoordinate.py:419]: Hi Ankit - this suggestion doesn't seem to have been applied - was that that intended?

File: cf/dimensioncoordinate.py (hunk 20 above)
Comment (davidhassell) [reply cf/dimensioncoordinate.py:343]: Hi Ankit - this, and other, suggestions are marked as reolved but are not reflected in the code - was that intended? Thanks.

**Review by davidhassell — CHANGES_REQUESTED** [review]
Hi Ankit - perhaps some git trouble again? Resolved suggestions don't seem to be reflected in the code. This is either because you disagreed with them, or the git thing is still troubling here ....

**Comment by bewithankit** [comment]
Hi David, I had applied the suggestions but have no idea why they are not reflected here! Probably due to the previous git trouble.

File: cf/dimensioncoordinate.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (bewithankit) [reply cf/dimensioncoordinate.py:418]: Hi David, you had suggested the following in https://github.com/NCAS-CMS/cf-python/pull/634#discussion_r1182885877:

> I realise that we don't need this, because the conversion to a `Units` object gets done for us inside the `Data` initialisation, below.

File: cf/dimensioncoordinate.py [hunk 27]
```diff
@@ -239,6 +239,110 @@ def direction(self):
...
+                "Range must be increasing for positive cellsizes."
+            )
+        elif cellsize < 0 and range_diff >= 0:
+            raise ValueError(
+                "Range must be decreasing for negative cellsizes."
+            )
+
+        if range_diff % cellsize != 0:
+            raise ValueError(
+                "cellsize must be divisible by the range of the dimension."
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:316]:
```suggestion
                f"The range of the dimension ({range_diff}) must be "
                f"divisible by the cellsize ({cellsize})"
```

File: cf/dimensioncoordinate.py [hunk 28]
```diff
@@ -239,6 +239,110 @@ def direction(self):
...
+                        f"Expected a sequence of three numbers, got {args}."
+                    )
+
+        range = (args[0], args[1])
+        cellsize = args[2]
+
+        range_diff = range[1] - range[0]
+        if cellsize > 0 and range_diff <= 0:
+            raise ValueError(
+                "Range must be increasing for positive cellsizes."
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:307]:
```suggestion
                f"Range ({range[0], range[1]}) must be increasing for a "
                f"positive cellsize ({cellsize})"
```

File: cf/dimensioncoordinate.py [hunk 29]
```diff
@@ -239,6 +239,110 @@ def direction(self):
...
+            )
+        elif cellsize < 0 and range_diff >= 0:
+            raise ValueError(
+                "Range must be decreasing for negative cellsizes."
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:311]:
```suggestion
                f"Range ({range[0], range[1]}) must be decreasing for a "
                f"negative cellsize ({cellsize})"
```

File: cf/test/test_Domain.py [hunk 30]
```diff
@@ -295,6 +295,52 @@ def test_Domain_transpose(self):
...
+        with self.assertRaises(ValueError):
+            cf.Domain.create_regular((-180, 180, 1), (-91, 90, 1))
+        with self.assertRaises(ValueError):
+            cf.Domain.create_regular((-180, 180, 1), (-90, 91, 1))
+
+        # Test for decreasing coordinates range
+        with self.assertRaises(ValueError):
+            cf.Domain.create_regular((180, -180, 1), (-90, 90, 1))
+        with self.assertRaises(ValueError):
+            cf.Domain.create_regular((-180, 180, 1), (90, -90, 1))
```
Comment (davidhassell) [inline cf/test/test_Domain.py:344]:
```suggestion
        # Test cyclicity
        d = cf.Domain.create_regular((-180, 180, 1), (90, -90, -1))
        axis = d.domain_axis('X', key=True)
        self.assertEqual(d.cyclic().pop(),axis)
```

File: cf/test/test_Domain.py [hunk 31]
```diff
@@ -295,6 +295,52 @@ def test_Domain_transpose(self):
     def test_Domain_size(self):
         self.assertEqual(self.d.size, 90)
 
+    def test_Domain_create_regular(self):
```
Comment (davidhassell) [inline cf/test/test_Domain.py:298]: We need a test for `bounds=False`.

File: cf/dimensioncoordinate.py [hunk 32]
```diff
@@ -239,6 +239,110 @@ def direction(self):
...
+        else:
+            start = range[0]
+            end = range[1]
+
+        points = np.arange(start, end + cellsize, cellsize)
+
+        coordinate = cls(
+            data = Data(points, units=units),
+            properties={"standard_name": standard_name},
+        )
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:338]:
```suggestion
            copy=False
        )
```

File: cf/dimensioncoordinate.py [hunk 33]
```diff
@@ -239,6 +239,110 @@ def direction(self):
...
+
+        if bounds:
+            b = coordinate.create_bounds()
+            coordinate.set_bounds(b)
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:342]:
```suggestion
            coordinate.set_bounds(b, copy=False)
```

File: cf/dimensioncoordinate.py [hunk 34]
```diff
@@ -239,6 +239,110 @@ def direction(self):
...
+
+            standard_name: str, optional
+                The standard_name of the `DimensionCoordinate` object.
+
+        :Returns:
+
+            `DimensionCoordinate`
+                The newly created DimensionCoordinate object.
+
+        **Examples:**
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:279]:
```suggestion
        **Examples**
```

**Review by davidhassell — CHANGES_REQUESTED** [review]
Hi Ankit - I'm really confused! The code I pull down  to test is not the same code as shown by the "Files Changed". I'm not sure how to proceed ....

File: cf/dimensioncoordinate.py [hunk 35]
```diff
@@ -239,6 +239,112 @@ def direction(self):
...
+    def create_regular(
+        cls, args, units=None, standard_name=None, bounds=True
+    ):
+        """
+        Create a new `DimensionCoordinate` with the given range and cellsize.
+
+        .. versionadded:: 3.15.0
+
+        :Note: This method does not set the cyclicity of the
+         `DimensionCoordinate`.
```
Comment (davidhassell) [inline cf/dimensioncoordinate.py:252]:
```suggestion
               `DimensionCoordinate`.
```

File: cf/test/test_DimensionCoordinate.py [hunk 36]
```diff
@@ -626,6 +626,36 @@ def test_DimensiconCoordinate_rechunk(self):
...
+            np.linspace(-179.5, 179.5, 360)))
+
+        # Invalid inputs
+        with self.assertRaises(ValueError):
+            cf.DimensionCoordinate.create_regular((-180, 180, 1, 2), 
+                units='degrees_east', standard_name='longitude')
+
+        with self.assertRaises(ValueError):
+            cf.DimensionCoordinate.create_regular((-180, 180, 1), units=123, 
+                standard_name='longitude')
```
Comment (davidhassell) [inline cf/test/test_DimensionCoordinate.py:644]:
```suggestion
```
This doesn't (and shouldn't) fail.

File: cf/domain.py [hunk 37]
```diff
@@ -330,6 +331,121 @@ def del_file_location(
...
+        domain_axis_longitude = domain.set_construct(
+            domain._DomainAxis(longitude.size)
+        )
+
+        domain_axis_latitude = domain.set_construct(
+            domain._DomainAxis(latitude.size)
+        )
+
+        domain.set_construct(longitude, axes=[domain_axis_longitude])
+        domain.set_construct(latitude, axes=[domain_axis_latitude])
```
Comment (davidhassell) [inline cf/domain.py:445]:
```suggestion
        domain.set_construct(longitude, axes=[domain_axis_longitude], copy=False)
        domain.set_construct(latitude, axes=[domain_axis_latitude], copy=False)
```

File: cf/domain.py [hunk 38]
```diff
@@ -330,6 +331,121 @@ def del_file_location(
...
+            x_args, "degrees_east", "longitude", bounds=bounds
+        )
+
+        latitude = domain._DimensionCoordinate.create_regular(
+            y_args, "degrees_north", "latitude", bounds=bounds
+        )
+
+        domain_axis_longitude = domain.set_construct(
+            domain._DomainAxis(longitude.size)
+        )
```
Comment (davidhassell) [inline cf/domain.py:438]:
```suggestion
            copy=False
        )
```

File: cf/domain.py [hunk 39]
```diff
@@ -330,6 +331,121 @@ def del_file_location(
...
+
+        domain_axis_latitude = domain.set_construct(
+            domain._DomainAxis(latitude.size)
+        )
```
Comment (davidhassell) [inline cf/domain.py:442]:
```suggestion
            copy=False
        )
```

File: cf/test/test_DimensionCoordinate.py [hunk 40]
```diff
@@ -626,6 +626,36 @@ def test_DimensiconCoordinate_rechunk(self):
...
         self.assertEqual(d.bounds.data.chunks, ((1,) * d.size, (2,)))
 
+    def test_DimensionCoordinate_create_regular(self):
+
+        longitude = cf.DimensionCoordinate.create_regular((-180, 180, 1), 
+            units='degrees_east', standard_name='longitude')
+        self.assertIsInstance(longitude, cf.DimensionCoordinate)
+        self.assertTrue(np.allclose(longitude.array, 
+            np.linspace(-179.5, 179.5, 360)))
+
```
Comment (davidhassell) [inline cf/test/test_DimensionCoordinate.py:636]:
```suggestion
        self.assertEqual(longitude.standard_name, "longitude")
        self.assertEqual(longitude.units, 'degrees_east')

```

File: cf/test/test_Domain.py [hunk 41]
```diff
@@ -295,7 +295,56 @@ def test_Domain_transpose(self):
...
+            cf.Domain.create_regular((-180, 180, 1), (-91, 90, 1))
+        with self.assertRaises(ValueError):
+            cf.Domain.create_regular((-180, 180, 1), (-90, 91, 1))
+
+        # Test for decreasing coordinates range
+        with self.assertRaises(ValueError):
+            cf.Domain.create_regular((180, -180, 1), (-90, 90, 1))
+        with self.assertRaises(ValueError):
+            cf.Domain.create_regular((-180, 180, 1), (90, -90, 1))
+        # Test cyclicity
```
Comment (davidhassell) [inline cf/test/test_Domain.py:344]:
```suggestion

        # Test cyclicity
```

**Review by davidhassell — CHANGES_REQUESTED** [review]
Hi Ankit - OK, I think we're looking good. Just a few more comments. Also, could you run the pre-commit linting?

File: cf/test/test_DimensionCoordinate.py [hunk 42]
```diff
@@ -626,6 +626,68 @@ def test_DimensiconCoordinate_rechunk(self):
...
+            np.allclose(longitude_decreasing.array, np.linspace(179.5, -179.5, 360))
+        )
+        self.assertEqual(longitude_decreasing.standard_name, "longitude")
+        self.assertEqual(longitude_decreasing.units, "degrees_east")
+
+        # Test decreasing case and bounds=False
+        longitude_decreasing_no_bounds = cf.DimensionCoordinate.create_regular(
+            (180, -180, -1), units="degrees_east", standard_name="longitude", bounds=False
+        )
+        self.assertIsInstance(longitude_decreasing_no_bounds, cf.DimensionCoordinate)
```
Comment (davidhassell) [inline cf/test/test_DimensionCoordinate.py:684]:
```suggestion
        self.assertIsInstance(longitude_decreasing_no_bounds, cf.DimensionCoordinate)
        self.assertFalse(longitude_decreasing_no_bounds.has_bounds())
```

File: cf/test/test_DimensionCoordinate.py [hunk 43]
```diff
@@ -626,6 +626,68 @@ def test_DimensiconCoordinate_rechunk(self):
...
+        longitude_decreasing = cf.DimensionCoordinate.create_regular(
+            (180, -180, -1), units="degrees_east", standard_name="longitude"
+        )
+        self.assertIsInstance(longitude_decreasing, cf.DimensionCoordinate)
+        self.assertTrue(
+            np.allclose(longitude_decreasing.array, np.linspace(179.5, -179.5, 360))
+        )
+        self.assertEqual(longitude_decreasing.standard_name, "longitude")
+        self.assertEqual(longitude_decreasing.units, "degrees_east")
+
```
Comment (davidhassell) [inline cf/test/test_DimensionCoordinate.py:679]:
```suggestion

        # Decreasing range with positive delta
        with self.assertRaises(ValueError):
            cf.DimensionCoordinate.create_regular((180, -180, 1))
            
```

File: cf/domain.py [hunk 44]
```diff
@@ -330,6 +331,123 @@ def del_file_location(
...
+        y_args = np.array(y_args)
+
+        if y_args.shape != (3,) or y_args.dtype.kind not in 'fi':
+            raise ValueError(
+                        "The args argument was incorrectly formatted. "
+                        f"Expected a sequence of three numbers, got {y_args}."
+                    )
+
+        x_range, dx = (x_args[0], x_args[1]), x_args[2]
+        y_range, dy = (y_args[0], y_args[1]), y_args[2]
```
Comment (davidhassell) [inline cf/domain.py:414]:
```suggestion
        x_range = x_args[0:2]
        y_range = y_args[0:2]
```

**Review by davidhassell — APPROVED** [review]
Hi Ankit - Nearly there!

A couple of new test suggestions and a bit of code tidy in `domain.py`.

Could you also run `pre-commit run --all-files` and commit those changes before merging? 

Thanks!
